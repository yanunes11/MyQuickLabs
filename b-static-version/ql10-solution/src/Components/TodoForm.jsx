import React from 'react';
import DateCreated from './utils/DateCreated';

const TodoForm = () => {

    return (
        <form>
            <div className="form-group">
                <label htmlFor="todoDescription">Description:&nbsp;</label>
                <input type="text" name="todoDescription" placeholder="Todo description" className="form-control" />
            </div>
            <div className="form-group">
                <label htmlFor="todoDateCreated">Created on:&nbsp;</label>
                <DateCreated />
            </div>
            <div className="form-group">
                <label htmlFor="todoCompleted">Completed:&nbsp;</label>
                <input type="checkbox" name="todoCompleted" />
            </div>
            <div className="form-group">
                <input type="submit" value="Submit" className="btn btn-primary" />
            </div>
        </form>
    );
};

export default TodoForm;
