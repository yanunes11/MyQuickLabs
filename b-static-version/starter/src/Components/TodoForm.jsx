import React from "react";
import DateCreated from './utils/DateCreated';
const TodoForm = () => {
    return (
        <>
            <form>
                <div className="form-group">
                    Todo Description:&nbsp;
                    <input className="form-control" type="text" placeholder="Todo Description"></input>
                </div>
                <div className="form-group" >
                    Todo Created:&nbsp;
                    <DateCreated/>
                </div>
                <div className="form-group" >
                    <input className="btn btn-primary" type="submit" value="Submit"></input>
                </div>
            </form>
        </>
    );
};

export default TodoForm;