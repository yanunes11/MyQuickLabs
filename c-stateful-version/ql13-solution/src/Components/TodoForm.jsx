import React, { useState } from 'react';
import DateCreated from './utils/DateCreated';
import PropTypes from 'prop-types';


const TodoForm = props => {

    const [todoDescription, setTodoDescription] = useState(``);
    const [todoDateCreated, setTodoDateCreated] = useState(null);
    const [todoCompleted, setTodoCompleted] = useState(false);

    const handleSubmit = event => {
        event.preventDefault();
        props.submitTodo(todoDescription, todoDateCreated, todoCompleted);
        setTodoDescription('');
        setTodoDateCreated(null);
        setTodoCompleted(false);
    }

    return (
        <form onSubmit={handleSubmit}>
            <div className="form-group">
                <label htmlFor="todoDescription">Description:&nbsp;</label>
                <input
                    type="text"
                    name="todoDescription"
                    placeholder="Todo description"
                    className="form-control"
                    value={todoDescription}
                    onChange={event => setTodoDescription(event.target.value)}
                />
            </div>
            <div className="form-group">
                <label htmlFor="todoDateCreated">Created on:&nbsp;</label>
                <DateCreated
                    updateDateCreated={dateCreated => setTodoDateCreated(dateCreated)}
                />
            </div>
            <div className="form-group">
                <label htmlFor="todoCompleted">Completed:&nbsp;</label>
                <input
                    type="checkbox"
                    name="todoCompleted"
                    checked={todoCompleted}
                    onChange={event => setTodoCompleted(event.target.checked)}
                />
            </div>
            <div className="form-group">
                <input type="submit" value="Submit" className={`btn ${!todoDescription ? `btn-danger` : `btn-primary`}`} disabled={!todoDescription} />
            </div>
        </form>
    );
};

TodoForm.propTypes = {
    submitTodo: PropTypes.func.isRequired
};

export default TodoForm;
