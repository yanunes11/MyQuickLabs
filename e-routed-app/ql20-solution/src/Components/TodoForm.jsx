import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import DateCreated from './utils/DateCreated';

const TodoForm = ({ submitAction, todo }) => {

    const [todoDescription, setTodoDescription] = useState(``);
    const [todoDateCreated, setTodoDateCreated] = useState(null);
    const [todoCompleted, setTodoCompleted] = useState(false);

    useEffect(() => {
        if (todo) {
            setTodoDescription(todo.todoDescription);
            setTodoDateCreated(todo.todoDateCreated);
            setTodoCompleted(todo.todoCompleted);
        }

        return (() => {
            setTodoDescription(``);
            setTodoDateCreated(null);
            setTodoCompleted(false);
        });
    }, [todo]);

    const handleSubmit = event => {
        event.preventDefault();
        submitAction(todoDescription, todoDateCreated, todoCompleted, todo?._id);
        setTodoDescription(``);
        setTodoDateCreated(null);
        setTodoCompleted(false);
    }

    return (
        <form onSubmit={handleSubmit}>
            <div className="form-group">
                <label htmlFor="todoDescription">Description:&nbsp;</label>
                <input
                    type="text"
                    name="todoDescription"
                    placeholder="Todo description"
                    className="form-control"
                    value={todoDescription}
                    onChange={event => setTodoDescription(event.target.value)}
                />
            </div>
            <div className="form-group">
                <label htmlFor="todoDateCreated">Created on:&nbsp;</label>
                {todo && `${new Date(todo.todoDateCreated).toLocaleDateString()} @ ${new Date(todo.todoDateCreated).toLocaleTimeString()}`}
                {!todo && <DateCreated updateDateCreated={dateCreated => setTodoDateCreated(dateCreated)} />}
            </div>
            { todo &&
                <div className="form-group">
                    <label htmlFor="todoCompleted">Completed:&nbsp;</label>
                    <input
                        type="checkbox"
                        name="todoCompleted"
                        checked={todoCompleted}
                        onChange={event => setTodoCompleted(event.target.checked)}
                    />
                </div>
            }
            <div className="form-group">
                <input type="submit" value="Submit" className={`btn ${!todoDescription ? `btn-danger` : `btn-primary`}`} disabled={!todoDescription} />
            </div>
        </form>
    );
};

TodoForm.propTypes = {
    submitAction: PropTypes.func.isRequired,
    todo: PropTypes.exact({
        todoDescription: PropTypes.string,
        todoDateCreated: PropTypes.string,
        todoCompleted: PropTypes.bool,
        _id: PropTypes.string
    })
}

export default TodoForm;
