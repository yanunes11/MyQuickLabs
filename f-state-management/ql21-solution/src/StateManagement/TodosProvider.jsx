import React, { createContext, useContext, useEffect, useState } from 'react';
import axios from 'axios';

export const TodosStateContext = createContext();
const baseUrl = `http://localhost:4000/todos`;

export const useTodosState = () => {
    const context = useContext(TodosStateContext);

    if (context === undefined) {
        throw new Error(`useTodosState must be used within a TodosProvider`);
    }

    return context;
}

const getAllTodos = async () => {
    try {
        const todosData = await axios.get(baseUrl);
        return todosData.data;
    }
    catch (error) {
        return { errorMessage: `Data not available from server: ${error.message}` };
    }
}

const TodosProvider = ({ children }) => {
    const [todos, setTodos] = useState({});

    useEffect(() => {
        const getTodos = async () => {
            const payload = await getAllTodos();
            setTodos(payload);
        };
        getTodos();
    }, []);

    return (
        <TodosStateContext.Provider value={{ todos }}>
            {children}
        </TodosStateContext.Provider>
    );
};

export default TodosProvider;