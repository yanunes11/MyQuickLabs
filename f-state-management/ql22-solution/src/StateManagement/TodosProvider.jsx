import React, { createContext, useContext, useEffect, /*useState*/ useReducer } from 'react';
import axios from 'axios';

export const TodosStateContext = createContext();
export const TodosDispatchContext = createContext();

const baseUrl = `http://localhost:4000/todos`;

const todosReducerAsyncHandler = dispatch => {
    let payload = {};
    return async action => {
        switch (action.type) {
            case `addTodo`:
                const addTodoResult = await addTodo(action.payload);
                payload = { ...addTodoResult, todos: await getAllTodos() }
                dispatch({ type: action.type, payload });
                break;
            case `editTodo`:
                const editTodoResult = await editTodo(action.payload);
                payload = { ...editTodoResult, todos: await getAllTodos() }
                dispatch({ type: action.type, payload });
                break;
            case `clearAddEditMessages`:
                dispatch({ type: action.type });
                break;
            default:
                dispatch({ type: `appError`, payload: { errorMessage: `There was an application error` } });
        }
    }
}

const todosReducer = (state, action) => {
    switch (action.type) {
        case `allTodos`:
            if (action.payload?.errorMessage) {
                return {
                    todos: null,
                    errors: { ...state.errors, get: action.payload.errorMessage }
                };
            }
            return {
                todos: action.payload,
                errors: null
            };
        case `addTodo`:
            if (action.payload.errorMessage) {
                return {
                    ...state,
                    errors: { ...state.errors, post: action.payload.errorMessage }, // Can I reduce this further?
                    success: null
                };
            }
            return {
                todos: action.payload.todos,
                errors: { ...state.errors, post: null },
                success: action.payload.successMessage
            };
        case `editTodo`:
            if (action.payload.errorMessage) {
                return {
                    ...state,
                    errors: { ...state.errors, put: action.payload.errorMessage },
                    success: null
                };
            }
            return {
                todos: action.payload.todos,
                errors: { ...state.errors, put: null },
                success: action.payload.successMessage
            };
        case `clearAddEditMessages`:
            return {
                ...state,
                errors: { ...state.errors, put: null, post: null },
                success: null
            };
        default:
            return {
                ...state,
                errors: { ...state.errors, [action.type]: action.payload.errorMessage },
                success: null
            };
    }
};

const getAllTodos = async () => {
    try {
        return (await axios.get(baseUrl)).data;
    }
    catch (error) {
        return { errorMessage: `Data not available from server: ${error.message}` };
    }
};

const addTodo = async newTodo => {
    try {
        await axios.post(baseUrl, newTodo);
        return { successMessage: `The todo was successfully added` };
    }
    catch (e) {
        return { errorMessage: `There was a problem adding the todo: ${e.message}` };
    }
};

const editTodo = async editedTodo => {
    try {
        await axios.put(`${baseUrl}/${editedTodo._id}`, editedTodo);
        return { successMessage: `The todo was successfully updated` };
    }
    catch (e) {
        return { errorMessage: `There was a problem updating the todo: ${e.message}` };
    }
}

export const useTodosState = () => {
    const context = useContext(TodosStateContext);

    if (context === undefined) {
        throw new Error(`useTodosState must be used within a TodosProvider`);
    }

    return context;
}

export const useTodosDispatch = () => {
    const context = useContext(TodosDispatchContext);

    if (context === undefined) {
        throw new Error(`useTodosDispatch must be used within a TodosProvider`);
    }

    return context;
}

const TodosProvider = ({ children }) => {
    const [state, dispatch] = useReducer(todosReducer, { todos: null, errors: {}, success: null });

    useEffect(() => {
        const getTodos = async () => {
            const payload = await getAllTodos();
            // setTodos(payload);
            dispatch({ type: `allTodos`, payload });
        };
        getTodos();
    }, []);

    return (
        <TodosStateContext.Provider value={state}>
            <TodosDispatchContext.Provider value={todosReducerAsyncHandler(dispatch)}>
                {children}
            </TodosDispatchContext.Provider>
        </TodosStateContext.Provider>
    );
};

export default TodosProvider;