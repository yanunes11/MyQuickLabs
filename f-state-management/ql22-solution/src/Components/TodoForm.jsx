import React, { useState, useEffect } from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import DateCreated from './utils/DateCreated';

import { useTodosDispatch, useTodosState } from '../StateManagement/TodosProvider';
import generateTodoId from './utils/generateId';
import TodoModel from './utils/Todo.model';
import Modal from './utils/Modal';

// const TodoForm = ({ submitAction, todo }) => {
const TodoForm = ({ todo }) => {

    const [todoDescription, setTodoDescription] = useState(``);
    const [todoDateCreated, setTodoDateCreated] = useState(null);
    const [todoCompleted, setTodoCompleted] = useState(false);
    const [submitResult, setSubmitResult] = useState();
    const dispatch = useTodosDispatch();
    const { errors, success } = useTodosState();

    useEffect(() => {
        if (todo) {
            setTodoDescription(todo.todoDescription);
            setTodoDateCreated(todo.todoDateCreated);
            setTodoCompleted(todo.todoCompleted);
        }

        return (() => {
            setTodoDescription(``);
            setTodoDateCreated(null);
            setTodoCompleted(false);
        });
    }, [todo]);

    useEffect(() => {
        if (errors?.post || errors?.put) {
            setSubmitResult({ message: errors.post ?? errors.put });
        }

        if (success) {
            setSubmitResult({ message: success });
        }

        return (() => setSubmitResult(null));
    }, [errors, success]);

    const createTodo = () => {
        const todoId = todo ? todo._id : generateTodoId();
        return new TodoModel(todoDescription, new Date(todoDateCreated).toISOString(), todoCompleted, todoId);
    }

    const handleSubmit = event => {
        event.preventDefault();
        // submitAction(todoDescription, todoDateCreated, todoCompleted, todo?._id);
        // setTodoDescription(``);
        // setTodoDateCreated(null);
        // setTodoCompleted(false);
        const type = todo ? `editTodo` : `addTodo`;
        const payload = createTodo();
        dispatch({ type, payload });
    }

    const closeModalHandler = () => {
        setSubmitResult({ ...submitResult, read: true });
        dispatch({ type: `clearAddEditMessages` });
    }

    return (
        <>
            { submitResult?.read && <Redirect to="/" />}
            { submitResult && !submitResult?.read && <Modal handleClose={closeModalHandler} message={submitResult.message} />}
            <form onSubmit={handleSubmit}>
                <div className="form-group">
                    <label htmlFor="todoDescription">Description:&nbsp;</label>
                    <input
                        type="text"
                        name="todoDescription"
                        placeholder="Todo description"
                        className="form-control"
                        value={todoDescription}
                        onChange={event => setTodoDescription(event.target.value)}
                    />
                </div>
                <div className="form-group">
                    <label htmlFor="todoDateCreated">Created on:&nbsp;</label>
                    {todo && `${new Date(todo.todoDateCreated).toLocaleDateString()} @ ${new Date(todo.todoDateCreated).toLocaleTimeString()}`}
                    {!todo && <DateCreated updateDateCreated={dateCreated => setTodoDateCreated(dateCreated)} />}
                </div>
                {todo &&
                    <div className="form-group">
                        <label htmlFor="todoCompleted">Completed:&nbsp;</label>
                        <input
                            type="checkbox"
                            name="todoCompleted"
                            checked={todoCompleted}
                            onChange={event => setTodoCompleted(event.target.checked)}
                            data-testid="completedCB"
                        />
                    </div>
                }
                <div className="form-group">
                    <input type="submit" value="Submit" className={`btn ${!todoDescription ? `btn-danger` : `btn-primary`}`} disabled={!todoDescription} />
                </div>
            </form>
        </>
    );
};

TodoForm.propTypes = {
    // submitAction: PropTypes.func.isRequired,
    todo: PropTypes.exact({
        todoDescription: PropTypes.string,
        todoDateCreated: PropTypes.string,
        todoCompleted: PropTypes.bool,
        _id: PropTypes.string
    })
}

export default TodoForm;
