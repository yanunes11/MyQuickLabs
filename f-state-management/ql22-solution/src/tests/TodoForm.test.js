import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, act, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import { customRenderStateAndDispatchContexts } from './__mocks__/testUtils';

import TodoForm from '../Components/TodoForm';
import sampleTodos from '../sampleTodos.json';
import axios from 'axios';
import TodosProvider from '../StateManagement/TodosProvider';

jest.mock("../Components/utils/DateCreated", () => {
    return function MockDateCreated() {
        return <span testid="dateCreated">Date Created Component</span>
    }
});

jest.mock("../Components/utils/generateId", () => (
    function generateTodoId() { return 1 }
));

jest.mock("axios");

describe(`TodoForm test suite`, () => {

    let providerProps;
    let dispatch;
    let dispatchProviderProps;

    describe(`Edit tests`, () => {

        beforeEach(() => {
            providerProps = { value: { todos: undefined, errors: {}, success: null } };
            dispatch = jest.fn();
            dispatchProviderProps = jest.fn(dispatch);
        });


        test(`should render the component initially`, async () => {
            let container;

            await act(async () => container = await customRenderStateAndDispatchContexts(<TodoForm />, providerProps, dispatchProviderProps));

            expect(container).toMatchSnapshot();
        });

        test('should have a todo already on the form if a todo is supplied in props', async () => {

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm todo={sampleTodos[0]} />, providerProps, dispatchProviderProps));

            const testDescription = sampleTodos[0].todoDescription;

            const todoDescription = await screen.findByDisplayValue(testDescription);

            expect(todoDescription.value).toBe(testDescription);

        });

        test('should have an enabled submit button if there is already a todo on the form', async () => {

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm todo={sampleTodos[0]} />, providerProps, dispatchProviderProps));

            const testDescription = sampleTodos[0].todoDescription;

            const todoDescription = await screen.findByDisplayValue(testDescription);
            const submitBtn = await screen.findByDisplayValue(/submit/i);

            expect(todoDescription.value).toBe(testDescription);
            expect(submitBtn.disabled).toBeFalsy();

        });

        test('should have a Completed checkbox if there is a todo supplied by props', async () => {

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm todo={sampleTodos[0]} />, providerProps, dispatchProviderProps));

            const completedCB = await screen.findByText(/completed/i);

            expect(completedCB).toBeInTheDocument();

        });

        test('should have a completed checkbox should change state when clicked', async () => {

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm todo={sampleTodos[3]} />, providerProps, dispatchProviderProps));

            const completedCB = await screen.findByTestId(/completedCB/i);

            expect(completedCB.checked).toBeFalsy();

            act(() => { userEvent.click(completedCB); });

            expect(completedCB.checked).toBeTruthy();

        });

        test('should display a modal with a success message when the submit button is clicked and a success response is given', async () => {
            axios.put.mockResolvedValueOnce({ data: sampleTodos[0] });

            providerProps = { value: { todos: sampleTodos, errors: {}, success: `The todo was successfully updated` } };
            dispatch = jest.fn();
            dispatchProviderProps = jest.fn(dispatch);

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm />, providerProps, dispatchProviderProps));

            const modalTitle = await screen.findByText(/todo application information/i);

            const successMessage = await screen.findByText(providerProps.value.success);

            expect(modalTitle).toBeInTheDocument();
            expect(successMessage).toBeInTheDocument();
        });

        test('should close the modal when close is clicked', async () => {

            providerProps = { value: { todos: sampleTodos, errors: {}, success: `The todo was successfully updated` } };
            dispatch = jest.fn();
            dispatchProviderProps = jest.fn(dispatch);

            await act(async () => await customRenderStateAndDispatchContexts(<MemoryRouter><TodoForm /></MemoryRouter>, providerProps, dispatchProviderProps));

            const modalClose = await screen.findByText(/close/i);

            await act(async () => await userEvent.click(modalClose));

            expect(modalClose).not.toBeInTheDocument();
        });

        test('should display an error modal when the server returns an error', async () => {
            providerProps = { value: { todos: sampleTodos, errors: { put: `put error` }, success: null } };
            dispatch = jest.fn();
            dispatchProviderProps = jest.fn(dispatch);

            await act(async () => await customRenderStateAndDispatchContexts(<MemoryRouter><TodoForm /></MemoryRouter>, providerProps, dispatchProviderProps));

            const modalTitle = await screen.findByText(/todo application information/i);

            const errorMessage = await screen.findByText(providerProps.value.errors.put);

            expect(modalTitle).toBeInTheDocument();
            expect(errorMessage).toBeInTheDocument();
        });


        test('should close the modal when close is clicked', async () => {

            providerProps = { value: { todos: sampleTodos, errors: { put: `put error` }, success: null } };
            dispatch = jest.fn();
            dispatchProviderProps = jest.fn(dispatch);

            await act(async () => await customRenderStateAndDispatchContexts(<MemoryRouter><TodoForm /></MemoryRouter>, providerProps, dispatchProviderProps));

            const modalClose = await screen.findByText(/close/i);

            await act(async () => await userEvent.click(modalClose));

            expect(modalClose).not.toBeInTheDocument();
        });
    });

    describe('Add Render tests', () => {

        test('should have an disabled submit button if there is no todo in props', async () => {

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm />, providerProps, dispatchProviderProps));

            const submitBtn = await screen.findByDisplayValue(/submit/i);

            expect(submitBtn.disabled).toBeTruthy();

        });

        test('should have an enabled submit button when the description box is typed into', async () => {

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm />, providerProps, dispatchProviderProps));

            const submitBtn = await screen.findByDisplayValue(/submit/i);

            expect(submitBtn.disabled).toBeTruthy();

            const todoDescInput = screen.getByPlaceholderText(/todo description/i);

            act(() => { userEvent.type(todoDescInput, 'Sample Todo'); });

            expect(submitBtn.disabled).toBeFalsy();
        });

    });

    describe('Submit Tests', () => {
        test('should call dispatch with correct action type and payload when editing', async () => {
            providerProps = { value: { todos: sampleTodos, errors: {}, success: null } };
            dispatch = jest.fn();
            dispatchProviderProps = jest.fn(dispatch);

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm todo={sampleTodos[3]} />, providerProps, dispatchProviderProps));

            const submitBtn = await screen.findByDisplayValue(/submit/i);

            act(() => { userEvent.click(submitBtn); });

            expect(dispatch).toHaveBeenCalledWith({ type: `editTodo`, payload: sampleTodos[3] });
        });

        test('should call dispatch with correct action type and payload when adding', async () => {
            providerProps = { value: { todos: sampleTodos, errors: {}, success: null } };
            dispatch = jest.fn();
            dispatchProviderProps = jest.fn(dispatch);

            await act(async () => await customRenderStateAndDispatchContexts(<TodoForm />, providerProps, dispatchProviderProps));

            const submitBtn = await screen.findByDisplayValue(/submit/i);
            const todoDescInput = screen.getByPlaceholderText(/todo description/i);

            act(() => { userEvent.type(todoDescInput, 'Sample Todo'); });
            act(() => { userEvent.click(submitBtn); });


            expect(dispatch).toHaveBeenCalled();
        });
    });
});

describe('tests for using the provider', () => {

    test('should render the form with a success modal when a todo is successfully added', async () => {
        axios.post.mockResolvedValue({ data: { todo: `Added` } });

        await act(async () => await render(<MemoryRouter><TodosProvider><TodoForm /></TodosProvider></MemoryRouter>));

        const submitBtn = await screen.findByDisplayValue(/submit/i);
        const todoDescInput = screen.getByPlaceholderText(/todo description/i);

        await act(async () => { await userEvent.type(todoDescInput, 'Sample Todo'); });
        await act(async () => { await userEvent.click(submitBtn); });

        const modalTitle = await screen.findByText(/todo application information/i);

        const successMessage = await screen.findByText(/the todo was successfully added/i);

        expect(modalTitle).toBeInTheDocument();
        expect(successMessage).toBeInTheDocument();
    });

    test('should render the form with an error modal when a todo is not added', async () => {
        axios.post.mockRejectedValue({ e: { message: `not added` } });

        await act(async () => await render(<MemoryRouter><TodosProvider><TodoForm /></TodosProvider></MemoryRouter>));

        const submitBtn = await screen.findByDisplayValue(/submit/i);
        const todoDescInput = screen.getByPlaceholderText(/todo description/i);

        await act(async () => { await userEvent.type(todoDescInput, 'Sample Todo'); });
        await act(async () => { await userEvent.click(submitBtn); });

        const modalTitle = await screen.findByText(/todo application information/i);

        const errorMessage = await screen.findByText(/There was a problem adding the todo/i);

        expect(modalTitle).toBeInTheDocument();
        expect(errorMessage).toBeInTheDocument();
    });

    test('should render the form with a success modal when a todo is successfully updated', async () => {
        axios.put.mockResolvedValue({ data: { todo: `Updated` } });

        await act(async () => await render(<MemoryRouter><TodosProvider><TodoForm todo={sampleTodos[3]} /></TodosProvider></MemoryRouter>));

        const submitBtn = await screen.findByDisplayValue(/submit/i);
        const todoDescInput = screen.getByPlaceholderText(/todo description/i);

        await act(async () => { await userEvent.click(submitBtn); });

        const modalTitle = await screen.findByText(/todo application information/i);

        const successMessage = await screen.findByText(/the todo was successfully updated/i);

        expect(modalTitle).toBeInTheDocument();
        expect(successMessage).toBeInTheDocument();
    });

    test('should render the form with an error modal when a todo is not updated', async () => {
        axios.put.mockRejectedValue({ e: { message: `not updated` } });

        await act(async () => await render(<MemoryRouter><TodosProvider><TodoForm todo={sampleTodos[3]} /></TodosProvider></MemoryRouter>));

        const submitBtn = await screen.findByDisplayValue(/submit/i);

        await act(async () => { await userEvent.click(submitBtn); });

        const modalTitle = await screen.findByText(/todo application information/i);

        const errorMessage = await screen.findByText(/There was a problem updating the todo/i);

        expect(modalTitle).toBeInTheDocument();
        expect(errorMessage).toBeInTheDocument();
    });

    test('should close the modal when Close is clicked, clearing any error messages', async () => {

        axios.put.mockRejectedValue({ e: { message: `not updated` } });

        await act(async () => await render(<MemoryRouter><TodosProvider><TodoForm todo={sampleTodos[3]} /></TodosProvider></MemoryRouter>));

        const submitBtn = await screen.findByDisplayValue(/submit/i);

        await act(async () => { await userEvent.click(submitBtn); });

        const modalTitle = await screen.findByText(/todo application information/i);

        const close = await screen.findByText(/close/i);

        await act(async () => await userEvent.click(close));

        expect(modalTitle).not.toBeInTheDocument();
        expect(close).not.toBeInTheDocument();

    })
});
