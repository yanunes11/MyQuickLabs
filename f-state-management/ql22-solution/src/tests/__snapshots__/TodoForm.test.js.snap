// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm test suite Edit tests should render the component initially 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="form-group"
        >
          <label
            for="todoDescription"
          >
            Description: 
          </label>
          <input
            class="form-control"
            name="todoDescription"
            placeholder="Todo description"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="todoDateCreated"
          >
            Created on: 
          </label>
          <span
            testid="dateCreated"
          >
            Date Created Component
          </span>
        </div>
        <div
          class="form-group"
        >
          <input
            class="btn btn-danger"
            disabled=""
            type="submit"
            value="Submit"
          />
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="form-group"
      >
        <label
          for="todoDescription"
        >
          Description: 
        </label>
        <input
          class="form-control"
          name="todoDescription"
          placeholder="Todo description"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="todoDateCreated"
        >
          Created on: 
        </label>
        <span
          testid="dateCreated"
        >
          Date Created Component
        </span>
      </div>
      <div
        class="form-group"
      >
        <input
          class="btn btn-danger"
          disabled=""
          type="submit"
          value="Submit"
        />
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
